torchaudio/prototype/rnnt_decoder.py

```diff
@@ -0,0 +1,342 @@
+from typing import Callable, Dict, List, Optional, NamedTuple, Tuple
+
+import torch
+
+from .rnnt import RNNT
+
+
+__all__ = ["RNNTBeamSearch"]
+
+
+class Hypo(NamedTuple):
+    tokens: List[int]
+    predictor_out: torch.Tensor
+    state: List[List[torch.Tensor]]
+    score: float
+    ali: List[int]
+    blank: int
+    key: str
+
+
+def _batch_state(hypos: List[Hypo]) -> List[List[torch.Tensor]]:
+    states: List[List[torch.Tensor]] = []
+    for i in range(len(hypos[0].state)):
+        batched_state_components: List[torch.Tensor] = []
+        for j in range(len(hypos[0].state[i])):
+            batched_state_components.append(
+                torch.cat([hypo.state[i][j] for hypo in hypos])
+            )
+        states.append(batched_state_components)
+    return states
+
+
+def _slice_state(
+    states: List[List[torch.Tensor]], idx: int, device: torch.device
+) -> List[List[torch.Tensor]]:
+    idx_tensor = torch.tensor([idx], device=device)
+    return [
+        [state.index_select(0, idx_tensor) for state in state_tuple]
+        for state_tuple in states
+    ]
+
+
+def _default_hypo_sort_key(hypo: Hypo) -> float:
+    return hypo.score / (len(hypo.tokens) + 1)
+
+
+def _compute_updated_scores(
+    hypos: List[Hypo], next_token_probs: torch.Tensor, beam_width: int,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    hypo_scores = torch.tensor([h.score for h in hypos]).unsqueeze(1)
+    nonblank_scores = (
+        hypo_scores + next_token_probs[:, :-1]
+    )  # [beam_width, num_tokens - 1]
+    nonblank_nbest_scores, nonblank_nbest_idx = nonblank_scores.reshape(-1).topk(
+        beam_width
+    )
+    nonblank_nbest_hypo_idx = nonblank_nbest_idx.div(
+        nonblank_scores.shape[1], rounding_mode="trunc"
+    )
+    nonblank_nbest_token = nonblank_nbest_idx % nonblank_scores.shape[1]
+    return nonblank_nbest_scores, nonblank_nbest_hypo_idx, nonblank_nbest_token
+
+
+def _remove_hypo(hypo: Hypo, hypo_list: List[Hypo]) -> None:
+    for i, elem in enumerate(hypo_list):
+        if hypo.key == elem.key:
+            del hypo_list[i]
+            break
+
+
+class RNNTBeamSearch(torch.nn.Module):
+    r"""Beam search decoder for RNN-T model.
+
+    Args:
+        model (RNNT): RNN-T model to use.
+        blank (int): index of blank token in vocabulary.
+        temperature (float, optional): temperature to apply to joint network output.
+            Larger values yield more uniform samples. (Default: 1.0)
+        hypo_sort_key (Callable[[Hypo], float], optional): callable that computes a score for a given
+            hypothesis to rank hypotheses by. (Default: _default_hypo_sort_key)
```

Comment:
1. Referring to private function in doctoring should be avoided.
2. I think it is better to explain the default behavior in the doctoring.

Therefore I suggest to use the signature, like `hypo_sort_key: Optional[Callable[[Hypo], float]] = None`, and explain the default sort behavior in plain expression. (I think it is average score?)